FHEM/10_SD_Rojaflex.pm

```diff
@@ -0,0 +1,639 @@
+				}
+				Log3 $name, 4, "$ioname: SD_Rojaflex set $name duration running time $duration s";
+				InternalTimer( (gettimeofday() + $duration), \&SD_Rojaflex_pctStop, $name );
+			}
+		} else {
+			$cmd = 'stop';
+		}
+	}
+
+	# Build msg and send it
+	my ($housecode,$channel) = split m/[_]/xms,$hash->{DEF};
+	my $msg = q(P109#08) . $housecode . sprintf('%X',$channel);
+	if ($cmd eq 'clearfav') {
+		$msg .= q(D); # gotofav
+	} else {
+		$msg .= $rev_codes{$cmd};
+	}
+	$msg .= q(A01);
+	if ($cmd eq 'clearfav') {
+		$msg .= q(D); # gotofav
+	} else {
+		$msg .= $rev_codes{$cmd};
+	}
+	$msg .= q(A);
+  my $sum = 0;
+  for (my $i = 7; $i < 20; $i += 2) {
+    $sum += hex(substr($msg, $i, 2));
+  }
+  $sum &= 0xFF;
+	$msg .= sprintf('%02X',$sum);
+	Log3 $name, 4, "$ioname: $name sendMsg=$msg";
+	for my $i (1 .. AttrVal($name, 'repetition', 1)) {
+		IOWrite($hash, 'sendMsg', $msg);
+	}
+
+	if ($cmd eq 'clearfav') {
+		my $timelongest = $timeToOpen;
+		if ($timeToClose > $timeToOpen) {$timelongest = $timeToClose};
+		Log3 $name, 4, "$ioname: SD_Rojaflex set $name clearFav running time $timelongest s";
+		InternalTimer( (gettimeofday() + $timelongest), \&SD_Rojaflex_clearfav, $name );
+		$hash->{clearfavcount} = 0;
+	}
+	# Calculate target position and motor state
+	if ($cmd eq 'down') {
+		if ($na == 1) {$tpos = '100'}; # nicht bei "set pct xx"
+		if ($cpos ne $tpos) {$motor = 'down'}; # Wenn nicht schon unten.
+		if ($cpos eq $tpos) {$motor = 'stop'}; # Wenn unten.
+	}
+	if ($cmd eq 'up') {
+		if ($na == 1) {$tpos = '0'}; # nicht bei "set pct xx"
+		if ($cpos ne $tpos) {$motor = 'up'}; # Wenn nicht schon oben.
+		if ($cpos eq $tpos) {$motor = 'stop'}; # Wenn oben.
+	}
+	if ($cmd eq 'stop' || $cmd eq 'savefav') {
+		$motor = 'stop';
+	}
+
+	$state = $cmd;
+	Log3 $name, 3, "$ioname: SD_Rojaflex set $name $state";
+
+	readingsBeginUpdate($hash);
+	readingsBulkUpdate($hash, 'state', $state, 1);
+	if ($state ne 'clearfav' && $state ne 'gotofav') {
+		# Wenn keine PositionUpdates vom Motor kommen, setze gleich die finale Position
+		if (AttrVal($name,'bidirectional',1) eq '0') {
+			# Jump direct to the final position, because we have no position updates and set motor stop
+			$cpos = $tpos;
+			$motor = 'stop';
+			# Save current position
+			if (AttrVal($name,'inversePosition',0) eq '1') {$cpos = 100 - $cpos}; # inverse position
+			readingsBulkUpdate($hash, 'pct', $cpos, 1);
+			readingsBulkUpdate($hash, 'cpos', $cpos, 1);
+		}
+		readingsBulkUpdate($hash, 'motor', $motor, 1);
+		if (AttrVal($name,'inversePosition',0) eq '1') {$tpos = 100 - $tpos}; # inverse position
+		readingsBulkUpdate($hash, 'tpos', $tpos, 1);
+	}
+	readingsEndUpdate($hash, 1);
+
+	# channel 0 set all devices, we must update all other devices with the same housecode
+	if ($channel eq '0') {
+		foreach my $d (keys %defs) {
+			if (defined($defs{$d}) && $defs{$d}{TYPE} eq 'SD_Rojaflex' && substr($defs{$d}{DEF},0,7) eq $housecode && substr($defs{$d}{DEF},-2) ne '_0') {
```

Comment:
Ich bin nicht sicher ob das komplett mit devspec2array ersetzbar wäre, aber in Teilen mindestens:

`devspec2array("TYPE=SD_Rojaflex")`  Damit hätte man schon mal alles vom Typ SD_Rojaflex :)

`devspec2array("TYPE=SD_Rojaflex:FILTER=DEF=^$housecode.*")`  Damit hätten wir zusäztlich alles was in DEF mit $housecode beginnt :)

Bisschen tricky wird das mit _0 am Ende, was nicht vorhanden sein soll. Ich vermute einfach, zweimal geht nicht zu filtern, also muss in die Regex rein, dass es nicht mit _0 endet:

`devspec2array("TYPE=SD_Rojaflex:FILTER=DEF=^$housecode(?!(_0$)).*$")`  

Testen konnte ich das nicht, weil mir dazu die passenden Defs fehlen und ich nicht ganz so durchblicke was es mit dem 0_ auf sich hat. 
So hätten wir aber eine Standard FHEM Filterung auf die passenden definitionen.

Reply:
Na prima, bei dem Versuch habe ich mir jetzt das komplette FHEM zerschossen :-(

Reply:
Einmal drüber schlafen, dann klappt es auch mit devspec2array :-)
Mit dem Regex hat es nicht funktioniert, aber man kann auch mehrere Filter angeben:

```
foreach my $d (devspec2array("TYPE=SD_Rojaflex:FILTER=DEF=$housecode.*:FILTER=DEF!=.*_0")) {
```

Reply:
Ah, mehrere Filter gehen doch, naja die Doku war nicht so ergiebig und ich hatte auf die schnelle kein Beispiel gefunden in dem das beschrieben ist.


Ich hab es so verstanden, dass es mit `_0 ` nicht aufhören darf, aber ein _01 erlaubt wäre. Ist das korrekt? Wenn ja, dann müsste auch auf das Ende in der Regex geprüft werden, wenn nein, dann ist das vermutlich auch der Fehler in meiner vorgeschlagenen Regex :)

$housecode würde ich auch vom Anfang weg suchen, geht schneller wenns da schon nicht passt.

`foreach my $d (devspec2array("TYPE=SD_Rojaflex:FILTER=DEF=^$housecode.*:FILTER=DEF!=.*_0$")) {`

Ich glaub ich muss mal selbst mit devspec2array rumprobieren, vermutlich kommt er mit dem $ nicht klar, weil nicht escaped :)

Reply:
Ich hab ein bisschen rumgespielt also 
Wenn die `$` escaped werden geht es, hier kannst Du es in der FHEM Kommandozeile testen, ansonsten versteht Perl das $ als Variable :(


{ join " -- ", devspec2array("TYPE=SD_Rojaflex:FILTER=DEF=^3122FD2(?!(_0\$)).*\$") }`` 


Ich habe für den Test drei Definitionen angelegt um es zu verifizieren:
`
SD_Rojaflex_3122FD2_0
SD_Rojaflex_3122FD2_01
SD_Rojaflex_3122FD2_9
`

Ein list 
`list TYPE=SD_Rojaflex:FILTER=DEF=^3122FD2(?!(_0$)).*$`

mit dem Filter liefert zwei Ergebnisse

`
SD_Rojaflex_3122FD2_01
SD_Rojaflex_3122FD2_9
`

Wenn ich das richtig interpretiert habe sollte auch nur einer weg gefiltert werden und zwar der, welcher mit ´_0` endet.
Wenn der mit _01 am Ende auch weg sollte, dann passt das $ am Ende der Regex nicht.

Reply:
Du hast wahrscheinlich Recht mit der Vermutung, das das $ am Ende nicht passt.
Das list funktioniert zwar, aber wenn ich das so einfüge:

`		foreach my $d (devspec2array("TYPE=SD_Rojaflex:FILTER=DEF=^3122FD2(?!(_0$)).*$")) {
`

erhalte ich beim "reload 10_SD_Rojaflex.pm" diese Ausgabe:

```
Final $ should be \$ or $name at ./FHEM/10_SD_Rojaflex.pm line 241, within string
syntax error at ./FHEM/10_SD_Rojaflex.pm line 241, near "("TYPE=SD_Rojaflex:FILTER=DEF=^3122FD2(?!(_0$)).*$""
Global symbol "$d" requires explicit package name (did you forget to declare "my $d"?) at ./FHEM/10_SD_Rojaflex.pm line 244.
Global symbol "$d" requires explicit package name (did you forget to declare "my $d"?) at ./FHEM/10_SD_Rojaflex.pm line 245.
Global symbol "$d" requires explicit package name (did you forget to declare "my $d"?) at ./FHEM/10_SD_Rojaflex.pm line 247.
Global symbol "$d" requires explicit package name (did you forget to declare "my $d"?) at ./FHEM/10_SD_Rojaflex.pm line 248.
Global symbol "$d" requires explicit package name (did you forget to declare "my $d"?) at ./FHEM/10_SD_Rojaflex.pm line 249.
Global symbol "$d" requires explicit package name (did you forget to declare "my $d"?) at ./FHEM/10_SD_Rojaflex.pm line 250.
Global symbol "$d" requires explicit package name (did you forget to declare "my $d"?) at ./FHEM/10_SD_Rojaflex.pm line 251.
Global symbol "$d" requires explicit package name (did you forget to declare "my $d"?) at ./FHEM/10_SD_Rojaflex.pm line 254.
syntax error at ./FHEM/10_SD_Rojaflex.pm line 258, near "}"
./FHEM/10_SD_Rojaflex.pm has too many errors.
```

Ich hatte dann gestern noch alle möglichen Varianten durchprobiert, das $ am Ende zu maskieren, aber erfolglos.

Es soll tatsächlich nur ein DEF mit "_0" gefiltert werden. Alle anderen Varianten von "_01" bis "_15" sollen bearbeitet werden.

Reply:
Es sind zwei ´$` in der Regex enthalten: Beide müssen maskiert werden. Dann klappts auch auf Perlebene. 

`foreach my $d (devspec2array("TYPE=SD_Rojaflex:FILTER=DEF=^3122FD2(?!(_0\$)).*\$")) {`


Das liegt auch nur daran, dass wir es in  "" gesetzt haben (doublequote) Ein Singlequote geht nicht, wegen dem Variablen housecode :(

Reply:
Was ich auch nicht wusste und wieder bemerkenswert finde ist, dass die Regex in devspec2array aufgepimpt wird und ein  `^` am Anfang und ein `\$` am Ende verpasst bekommt. Jetzt verstehe ich auch, wieso das bei dir geklappt hat mit dem 2. Filter.

Reply:
Das zweite "$" hatte ich tatsächlich nicht beachtet - ich habe mich wohl zu sehr auf die Fehlermeldung konzentriert: "Final $ should be \$ or $name" :-(
Jetzt funktioniert es offensichtlich auch mit diesem Regex.

@Hofyyy 
Machst du bitte auch immer mal ein Update zwischendurch auf diesen Branch. Ich gebe mir zwar größte Mühe, alle Änderungen, die jetzt noch anfallen durchzutesten, aber in der Praxis sieht es vielleicht doch anders aus.